exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/map/UncheckedFlatMapProof.java

```diff
@@ -249,39 +258,55 @@ private CheckedMapProof checkProof() {
     return commonPrefix;
   }
 
-  private static HashCode getEmptyProofListHash() {
-    return HashCode.fromBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES]);
+  private static HashCode getEmptyProofIndexHash() {
+    HashCode merkleRoot = HashCode.fromBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES]);
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(MapProofEntry entry) {
-    return getSingleEntryRootHash(entry.getDbKey(), entry.getHash());
+  private static HashCode getSingleEntryProofIndexHash(MapProofEntry proofEntry) {
+    HashCode merkleRoot = getSingleEntryMerkleRoot(proofEntry.getDbKey(),
+        proofEntry.getHash());
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(MapEntry<ByteString, ByteString> entry) {
-    DbKey dbKey = DbKey.newLeafKey(entry.getKey());
-    HashCode valueHash = HASH_FUNCTION.hashByteString(entry.getValue());
-    return getSingleEntryRootHash(dbKey, valueHash);
+  private static HashCode getSingleEntryProofIndexHash(MapEntry<ByteString, ByteString> mapEntry) {
+    DbKey dbKey = DbKey.newLeafKey(mapEntry.getKey());
+    HashCode valueHash = getLeafEntryHash(mapEntry.getValue());
+    HashCode merkleRoot = getSingleEntryMerkleRoot(dbKey, valueHash);
+    return getIndexHash(merkleRoot);
   }
 
-  private static HashCode getSingleEntryRootHash(DbKey key, HashCode valueHash) {
+  private static HashCode getIndexHash(HashCode merkleRoot) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(MAP_ROOT_PREFIX)
+        .putObject(merkleRoot, hashCodeFunnel())
+        .hash();
+  }
+
+  private static HashCode getSingleEntryMerkleRoot(DbKey key, HashCode valueHash) {
     assert key.getNodeType() == Type.LEAF;
     return HASH_FUNCTION.newHasher()
+        .putByte(MAP_NODE_PREFIX)
         .putObject(key, dbKeyFunnel())
```

Comment:
Is that correct that it does not compress here? If yes, is this inconsistency worth it (in the core and all the clients), is it somehow justified?

Reply:
Yes, I also noticed this inconsistency, but it's like that in core - https://github.com/exonum/exonum/blob/master/components/merkledb/src/hash.rs#L143
@pavel-mukhanov is there a reason why database key path isn't compressed in case of a single entry proof?